src/basic/hashmap.c

```diff
@@ -1976,3 +1976,36 @@ IteratedCache* iterated_cache_free(IteratedCache *cache) {
 
         return mfree(cache);
 }
+
+int set_strjoin(Set *s, const char *separator, char **ret) {
+        size_t separator_len, allocated = 0, len = 0;
```

Comment:
It doesn't matter too much, but right now the end position is accounted twice: once in `len` and once in `e`. I think it'd be cleaner to just use one variable:
```
        SET_FOREACH(value, s) {
                size_t l = strlen_ptr(value);

                if (l == 0)
                          continue;

                if (!GREEDY_REALLOC(str, allocated, len + l + (first ? 0 : separator_len) + 1))
                        return -ENOMEM;

                if (separator_len > 0 && !first)
                        memcpy(str + len, separator, separator_len);
                memcpy(str + len + first ? 0 : separator_len, value, l);
                len += (first ? 0 : separator_len) + l;
                first = false;
        }
        str[len] = '\0';
```

This has the advantage that we use the slightly faster memcpy instead of strcpy too.